File: FordFulkerson.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

/**
 * @author ChatGPT, Mopheshi
 */
public class FordFulkerson {

    private static final int V = 6; // Number of vertices in the graph

    /**
     * Returns true if there is a path from source 's' to sink 't' in residual
     * graph. Also fills parent[] to store the path
     */
    private static boolean bfs(int[][] residualGraph, int source, int sink, int[] parent) {
        boolean[] visited = new boolean[V];
        Arrays.fill(visited, false);

        // Create a queue for BFS
        Queue<Integer> queue = new LinkedList<>();
        queue.add(source);
        visited[source] = true;
        parent[source] = -1;

        while (!queue.isEmpty()) {
            int u = queue.poll();

            for (int v = 0; v < V; v++) {
                if (!visited[v] && residualGraph[u][v] > 0) {
                    queue.add(v);
                    parent[v] = u;
                    visited[v] = true;
                }
            }
        }

        return visited[sink];
    }

    /**
     * A function to implement the Ford-Fulkerson algorithm
     *
     * @param graph
     * @param source
     * @param sink
     * @return maxFlow
     */
    public static int fordFulkerson(int[][] graph, int source, int sink) {
        // Create a residual graph and fill the residual graph with the given capacities
        int[][] residualGraph = new int[V][V];
        for (int i = 0; i < V; i++) {
            // Copies an array from the specified source array, beginning at the
            // specified position, to the specified position of the destination array.
            System.arraycopy(graph[i], 0, residualGraph[i], 0, V);
        }

        int[] parent = new int[V]; // This array is filled by BFS and to store the path

        int maxFlow = 0; // There is no flow initially

        // Augment the flow while there is a path from source to sink
        while (bfs(residualGraph, source, sink, parent)) {
            // Find the maximum flow through the path found by BFS
            int pathFlow = Integer.MAX_VALUE;
            for (int v = sink; v != source; v = parent[v]) {
                int u = parent[v];
                pathFlow = Math.min(pathFlow, residualGraph[u][v]);
            }

            // Update residual capacities and reverse edges along the path
            for (int v = sink; v != source; v = parent[v]) {
                int u = parent[v];
                residualGraph[u][v] -= pathFlow;
                residualGraph[v][u] += pathFlow;
            }

            // Add path flow to the overall maximum flow
            maxFlow += pathFlow;
        }

        // Return the overall maximum flow
        return maxFlow;
    }

    /**
     * Driver code in main method
     *
     * @param args
     */
    public static void main(String[] args) {
        // Create a graph with capacity matrix
        int[][] graph = {
                {0, 16, 13, 0, 0, 0},
                {0, 0, 10, 12, 0, 0},
                {0, 4, 0, 0, 14, 0},
                {0, 0, 9, 0, 0, 20},
                {0, 0, 0, 7, 0, 4},
                {0, 0, 0, 0, 0, 0}
        };

        int source = 0; // Source vertex
        int sink = 5; // Sink vertex

        int maxFlow = fordFulkerson(graph, source, sink);
        System.out.printf("The maximum possible flow is %d%n", maxFlow);
    }
}

File: ford_fulkerson.py
from collections import defaultdict

"""
Implementation of the Ford-Fulkerson algorithm.
"""


def bfs(graph, source, sink, parent):
    """
    Breadth-First Search algorithm
    :param graph:
    :param source:
    :param sink:
    :param parent:
    :return boolean:
    """
    visited = set()
    queue = [(source, float('inf'))]
    visited.add(source)

    while queue:
        u, flow = queue.pop(0)
        for v, capacity in graph[u]:
            if v not in visited and capacity > 0:
                visited.add(v)
                parent[v] = (u, capacity)
                if v == sink:
                    return True
                queue.append((v, min(flow, capacity)))
    return False


def ford_fulkerson(graph, source, sink):
    """
    Ford-Fulkerson algorithm.
    :param graph:
    :param source:
    :param sink:
    :return maximum flow:
    """
    parent = {}
    max_flow = 0

    while bfs(graph, source, sink, parent):
        path_flow = float('inf')
        s = sink
        while s != source:
            path_flow = min(path_flow, parent[s][1])
            s = parent[s][0]

        max_flow += path_flow
        v = sink
        while v != source:
            u = parent[v][0]
            graph[u][v] -= path_flow
            graph[v][u] += path_flow
            v = parent[v][0]

    return max_flow


# Implementation to find the maximum flow through a network:
graph = defaultdict(dict)
graph[0][1] = 16
graph[0][2] = 13
graph[1][2] = 10
graph[1][3] = 12
graph[2][1] = 4
graph[2][4] = 14
graph[3][2] = 9
graph[3][5] = 20
graph[4][3] = 7
graph[4][5] = 4

source = 0
sink = 5

max_flow = ford_fulkerson(graph, source, sink)
print("Maximum Flow:", max_flow)
